ships/0025-strategy-parameters-enhancements.md

```diff
@@ -0,0 +1,282 @@
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+None
+
+## Summary
+
+Recently, we introduced build strategy parameters with default values support through [ship 0014](0014-parameterize-strategies.md). This only supported single strings as values.
+
+This ship proposes an extension to support arrays, Secret and ConfigMap values for strategy parameters.
+
+## Motivation
+
+Build strategies need their own way to define parameters to enable certain features. In certain scenarios, single strings are not enough for all purposes:
+
+- There can be many [ARGs in a Dockerfile](https://docs.docker.com/engine/reference/builder/#using-arg-variables) for which the corresponding value needs to be provided when running the build.
+- Values of a build-arg are sometimes secrets, for example to pass a value for the [NPM_TOKEN environment variable](https://docs.npmjs.com/using-private-packages-in-a-ci-cd-workflow) into a Dockerfile build. Users do not want to provide those through a text field in a Build or BuildRun but rather using a Secret key reference.
+- Values of a build-arg are sometimes common across builds, for example the version of a tool that is used across builds, for example the version of a base image. Users do not want to update several builds individually, potentially forgetting one and continuing to build with an outdated version. Instead they want re-usable values stored in ConfigMaps and reference its keys in multiple Builds or BuildRuns.
+
+That's where the need for the additional features comes from.
+
+### Goals
+
+- Allow the build strategy author to define if a parameter is a single value or an array.
+- Allow the build strategy author to define a default value also for an array parameter.
+- Provide updated build strategy samples that use arrays so that the build strategy author can understand how to do this.
+- Allow build users to define values for an array parameter using simple strings.
+- Allow build users to define values for a string parameter, or an element of an array parameter, using a Secret key reference.
+- Allow build users to define values for a string parameter, or an element of an array parameter, using a ConfigMap key reference.
+- Allow build users to define a prefix for the Secret value.
+- Allow build users to define a prefix for a ConfigMap value.
+
+### Non-Goals
+
+## Proposal
+
+### API Changes
+
+We will extend the Build Strategy Parameter type to support the following:
+
+```yaml
+spec:
+  parameters:
+    - name: build-args
+      description: The values for the ARGs in the Dockerfile.
+      type: array
+      defaults: []
+```
+
+The `type` property is new and allows either `string` or `array` with `string` being the default.
+
+The `defaults` parameter is new and is a string array. No possibility here to reference Secret or ConfigMap values.
```

Comment:
> No possibility here to reference Secret or ConfigMap values.
But it kind of would be nice to have.

Reply:
Well, not for cluster build strategies in my opinion because they are cluster-scoped while secrets and configmaps are namespaced. One could do it just for namespaced build strategies at some point.